### PR TITLE
Deduplicate libraries on hash instead of filename.

### PR DESCRIPTION
**Pull request by taralx**
Removes the need for canonicalization to prevent #12459.

(Now with passing tests!)

Canonicalization breaks certain environments where the libraries are symlinks to files that don't end in .rlib (e.g. /remote/cas/$HASH).

### CODE REVIEW

**Comment by nagisa** [comment]
r? @alexcrichton 

hfive ain’t workin

**Comment by alexcrichton** [comment]
Thanks for the PR @taralx! Some of the test changes here are a little worrisome to me, does this break existing tests? If not, could this just add a new standalone test (or append to an existing one?)

I'd also want to think a little bit on this as well, I feel like we've intentionally not done this in the past but I can't really see any reason *not* to do this when I think about it...

**Comment by taralx** [comment]
Yes, this breaks existing tests that assume that two libraries with the same name but identical content will not be deduplicated.

File: src/test/run-make/compiler-lookup-paths/Makefile [hunk 1]
```diff
@@ -21,7 +21,7 @@ all: $(TMPDIR)/libnative.a
 	mkdir -p $(TMPDIR)/e1
 	mkdir -p $(TMPDIR)/e2
 	$(RUSTC) e.rs -o $(TMPDIR)/e1/libe.rlib
-	$(RUSTC) e.rs -o $(TMPDIR)/e2/libe.rlib
+	$(RUSTC) e.rs -o $(TMPDIR)/e2/libe-alt.rlib
```
Comment (taralx) [inline src/test/run-make/compiler-lookup-paths/Makefile]: Changed because two libraries with the same name and content but different paths will not conflict anymore.

File: src/test/run-make/extern-flag-fun/Makefile [hunk 2]
```diff
@@ -3,14 +3,15 @@
...
 	$(RUSTC) bar.rs --crate-type=rlib
 	$(RUSTC) bar.rs --crate-type=rlib -C extra-filename=-a
+	$(RUSTC) bar-alt.rs --crate-type=rlib
 	$(RUSTC) foo.rs --extern hello && exit 1 || exit 0
 	$(RUSTC) foo.rs --extern bar=no-exist && exit 1 || exit 0
 	$(RUSTC) foo.rs --extern bar=foo.rs && exit 1 || exit 0
 	$(RUSTC) foo.rs \
 		--extern bar=$(TMPDIR)/libbar.rlib \
-		--extern bar=$(TMPDIR)/libbar-a.rlib \
+		--extern bar=$(TMPDIR)/libbar-alt.rlib \
```
Comment (taralx) [inline src/test/run-make/extern-flag-fun/Makefile]: Changed because two externs with the same hash don't conflict anymore.

File: src/test/run-make/extern-flag-fun/Makefile [hunk 3]
```diff
@@ -3,14 +3,15 @@
...
 		&& exit 1 || exit 0
 	$(RUSTC) foo.rs \
 		--extern bar=$(TMPDIR)/libbar.rlib \
-		--extern bar=$(TMPDIR)/libbar.rlib
+		--extern bar=$(TMPDIR)/libbar-a.rlib
```
Comment (taralx) [inline src/test/run-make/extern-flag-fun/Makefile]: Enhanced to test that deduplication is based on hash and not only name.

File: src/test/run-make/issue-11908/Makefile [hunk 4]
```diff
@@ -12,12 +12,11 @@ all:
 	$(RUSTC) foo.rs --crate-type=dylib
 	mv $(call DYLIB,foo) $(TMPDIR)/other
 	$(RUSTC) foo.rs --crate-type=dylib
-	$(RUSTC) bar.rs -L $(TMPDIR)/other 2>&1 | \
-		grep "multiple dylib candidates"
+	# dylibs fail with strange errors, but not assertions.
+	! $(RUSTC) bar.rs -L $(TMPDIR)/other 2>&1 | grep "assertion failed"
```
Comment (taralx) [inline src/test/run-make/issue-11908/Makefile]: As per the comment -- this doesn't fail with "multiple dylib candidates" anymore (the dylibs have the same hash), but still fails for reasons I do not understand.

File: src/test/run-make/issue-11908/Makefile [hunk 5]
```diff
@@ -12,12 +12,11 @@ all:
...
 	rm -rf $(TMPDIR)
 	mkdir -p $(TMPDIR)/other
 	$(RUSTC) foo.rs --crate-type=rlib
 	mv $(TMPDIR)/libfoo.rlib $(TMPDIR)/other
 	$(RUSTC) foo.rs --crate-type=rlib
-	$(RUSTC) bar.rs -L $(TMPDIR)/other 2>&1 | \
-		grep "multiple rlib candidates"
+	$(RUSTC) bar.rs -L $(TMPDIR)/other
```
Comment (taralx) [inline src/test/run-make/issue-11908/Makefile]: This is the primary purpose of this change -- this should not fail as the two crates are basically the same.

**Comment by alexcrichton** [comment]
@taralx could this be changed to not modify existing tests where possible? If we want to add tests for new functionality then new tests should probably be added rather than tweaking old ones

File: src/test/run-make/issue-11908/Makefile (hunk 4 above)
Comment (alexcrichton) [inline src/test/run-make/issue-11908/Makefile]: This seems pretty scary, how come this is triggering an assertion?

Comment (taralx) [inline src/test/run-make/issue-11908/Makefile]: It's actually *not* triggering an assertion, which is what issue 11908 was about. The ! in the front inverts the return of the entire pipe.

The actual output is:

error: cannot satisfy dependencies so `std` only shows up once
help: having upstream crates all available in one format will likely make this go away
error: cannot satisfy dependencies so `core` only shows up once
help: having upstream crates all available in one format will likely make this go away
error: cannot satisfy dependencies so `collections` only shows up once
help: having upstream crates all available in one format will likely make this go away
error: cannot satisfy dependencies so `rustc_unicode` only shows up once
help: having upstream crates all available in one format will likely make this go away
error: cannot satisfy dependencies so `alloc` only shows up once
help: having upstream crates all available in one format will likely make this go away
error: cannot satisfy dependencies so `rand` only shows up once
help: having upstream crates all available in one format will likely make this go away
error: cannot satisfy dependencies so `libc` only shows up once
help: having upstream crates all available in one format will likely make this go away
error: cannot link together two allocators: alloc_jemalloc and alloc_system
error: aborting due to 8 previous errors

**Comment by taralx** [comment]
@alexcrichton I've made the absolutely minimal changes to existing tests, as noted above in line notes. The problem is that various tests were making assumptions about our behavior with multiple identical crates, which is not a good way to test things. Where possible, I've changed it so that the crates are either (a) not identical or (b) differently named.

**Comment by taralx** [comment]
The other alternative, if you prefer it, is to rip out the fragile tests and make new ones that subsume them and more clearly cover the space.

File: src/test/run-make/compiler-lookup-paths/Makefile (hunk 1 above)
Comment (alexcrichton) [inline src/test/run-make/compiler-lookup-paths/Makefile]: Hm I'm a little confused here, don't both these libraries have the same SVH? I don't follow why changing the filename causes them to suddenly be "different"

File: src/test/run-make/extern-flag-fun/Makefile (hunk 2 above)
Comment (alexcrichton) [inline src/test/run-make/extern-flag-fun/Makefile]: This seems like it should still be an error? Passing `--extern bar=...` to two *distinct* locations seems like it's surely an error?

File: src/test/run-make/issue-11908/Makefile (hunk 4 above)
Comment (alexcrichton) [inline src/test/run-make/issue-11908/Makefile]: This is really worrisome to me, this should generate the same error message it seems like? Either that or the two dylibs here need to have different hashes to generate the same error.

File: src/test/run-make/issue-11908/Makefile (hunk 5 above)
Comment (alexcrichton) [inline src/test/run-make/issue-11908/Makefile]: Can you change this test so it still exhibits the same error message?

**Comment by alexcrichton** [comment]
I'd rather keep the tests continue testing the same properties if possible, but can you also add some tests specifically for this PR itself?

File: src/test/run-make/compiler-lookup-paths/Makefile (hunk 1 above)
Comment (taralx) [inline src/test/run-make/compiler-lookup-paths/Makefile]: same SVH, different "hash" i.e. the part of the filename between lib and .rlib.

File: src/test/run-make/extern-flag-fun/Makefile (hunk 2 above)
Comment (taralx) [inline src/test/run-make/extern-flag-fun/Makefile]: Externs are processed as a single group, so we deduplicate on hash irrespective of the filename.

File: src/test/run-make/issue-11908/Makefile (hunk 4 above)
Comment (taralx) [inline src/test/run-make/issue-11908/Makefile]: It *should* generate no error, but I'm in no way equipped to debug the error. It doesn't seem related to my change though -- even if you only had one copy this would fail.

**Comment by taralx** [comment]
So I'll grant that the current situation is kind of confusing, because we have two levels of duplicate checking:

1. For libraries found by search, we first group them by basename, then extract one from each group (which complains if we see multiple hashes). If more than one group returns a library, we complain.
2. For externs (and other direct searches?), we simply group all of the libraries together and extract one from the single group. So we'll only complain if we hash multiple hashes, regardless of basename.

What is the desired ruleset?

**Comment by alexcrichton** [comment]
Hm there may be a disconnect about "hash" here that we've been talking about, and also perhaps some discrepancies in the implementation. The "hash" in the filename, e.g. `libfoo-xxxx.rlib` where the hash is "xxxx", isn't actually a hash at all. It's just some disambiguating information so multiple libraries of the same crate name can sit on the filesystem. Duplicate hashes there don't actually mean anything as they don't indicate anything about the contents.

Inside the libraries, however, there's a hash called the SVH (strict version hash) which is indeed a hash of the contents of the libraries. We should be deduplicating based on the SVH, not the filename. Does that make sense?

**Comment by taralx** [comment]
Okay, I think I know what's going on. I think that once upon a time, we used to filter on filename *first*, then only read metadata if we needed to disambiguate further. Now, we read metadata and ignore filename entirely.

I'm going to push this all the way to metadata only with SVH-based dedup, ignoring filenames completely (except for .rlib suffix).

File: src/test/run-make/issue-11908/Makefile (hunk 5 above)
Comment (taralx) [inline src/test/run-make/issue-11908/Makefile]: I don't think that's a good idea. Issue 11908 is about getting assertion failures, not a missing check for "multiple rlib candidates". I will change it to explicitly test for not "assertion failed", which is what the issue was about.

**Comment by taralx** [comment]
So it turns out grouping by name avoid matching an rlib with a different dylib (e.g. liba.rlib with libarena.dylib when searching for "a").

Still trying to find a solution.

**Comment by taralx** [comment]
@alexcrichton Ok! Have another look, hopefully the test diffs are more understandable.

**Comment by taralx** [comment]
I have one more little change to remove the last calls to fs::canonicalize coming as soon as I check tests.

File: src/test/run-make/issue-11908/Makefile [hunk 6]
```diff
@@ -13,11 +13,11 @@ all:
 	mv $(call DYLIB,foo) $(TMPDIR)/other
 	$(RUSTC) foo.rs --crate-type=dylib
 	$(RUSTC) bar.rs -L $(TMPDIR)/other 2>&1 | \
-		grep "multiple dylib candidates"
+		grep "assertion failed" && exit 1 || exit 0
```
Comment (alexcrichton) [inline src/test/run-make/issue-11908/Makefile]: I think these tests should pass, right? With your changes the SVH is the same so they should get deduplicated so the compilation should succeed?

I'm a little wary to have a test that the compiler doesn't print "assertion failed" as that's a *very* brittle test to easily break without actually detecting a regression, but if these compiles are successful maybe that part can be tested?

Comment (taralx) [inline src/test/run-make/issue-11908/Makefile]: No, the dylib one doesn't pass -- it prints other strange errors. The issue in question was "this causes the compiler to issue assertion failures". We could extend it to "these should work", but then we'd need to fix the other issue with dylibs.

Comment (alexcrichton) [inline src/test/run-make/issue-11908/Makefile]: What are the "strange errors" that get printed here? 

"The compiler doesn't assert" is such a broad test case it's essentially not testing anything unfortunately.

Comment (taralx) [inline src/test/run-make/issue-11908/Makefile]: https://github.com/rust-lang/rust/pull/32317#discussion_r57373547

Comment (alexcrichton) [inline src/test/run-make/issue-11908/Makefile]: Ok to take care of those you can just compile all dylibs here with `-C prefer-dynamic`

Comment (taralx) [inline src/test/run-make/issue-11908/Makefile]: Done. :)

**Comment by alexcrichton** [comment]
Looks good to me, thanks! Can you also squash the commits?

**Comment by taralx** [comment]
h5 has previously said not to do that preemptively.

Squash and rebase, or just squash?

**Comment by alexcrichton** [comment]
Ah yeah that was mainly intended for just review, once it's ready to go we prefer to squash + rebase before landing.

**Comment by taralx** [comment]
Done. Assuming tests pass, this one's good to go.

**Comment by alexcrichton** [comment]
@bors: r+ 393b86a8071980f3f89205a44653a35bd7a269b1

Thanks!

**Comment by taralx** [comment]
That's a weird failure. Does this happen often?

**Comment by alexcrichton** [comment]
Looks like an ICE:

```
thread 'rustc' panicked at 'we should have found local duplicate earlier', C:\bot\slave\auto-win-msvc-64-opt-rustbuild\build\src\libcore\option.rs:700
```

definitely seems related to this patch though?

**Comment by taralx** [comment]
Possibly. It apparently only shows up during rustbuild, and the error unfortunately doesn't include the lang item. :-/

**Comment by taralx** [comment]
Hm, hm. No local repro. I can compile std via cargo just fine. Maybe retry? I can't see a reason why this would change things in a normal build.

**Comment by taralx** [comment]
Maybe I need to --enable-debug

**Comment by alexcrichton** [comment]
Ah yeah without that I think assertions may be disabled by default?

**Comment by taralx** [comment]
No repro with --enable-debug.

**Comment by alexcrichton** [comment]
I think the incantation may be `--enable-debug-assertions`?

**Comment by taralx** [comment]
--enable-debug --enable-debug-assertion no repro :(

**Comment by alexcrichton** [comment]
Hm ok, retry with a rebase?

**Comment by taralx** [comment]
Okay, let's try it.

**Comment by taralx** [comment]
Seems strangely restricted to windows platforms, which build differently in rustbuild. Any idea on how I can repro this for debugging, or what might be causing it?

**Comment by taralx** [comment]
Is it doing a parallel compile, or did the error show up at a different place this time?

**Comment by alexcrichton** [comment]
Crates are indeed compiling in parallel, and looks like this is legit?

**Comment by taralx** [comment]
Yeah, definitely legit, but I can't imagine what's causing it. It may be necessary to add more diagnostics to the build.

**Comment by taralx** [comment]
Is there a way to run the windows build manually somehow? I don't have a box with the necessary compiler on it myself. Maybe a cloud instance?

**Comment by alexcrichton** [comment]
Nah unfortunately we don't have a great infrastructure for doing so :(

**Comment by taralx** [comment]
So it turns out that creader was also calling fs::canonicalize during existing_match under the assumption that all paths are canonical. Instead of removing it, I just backed out that change, leaving everything using canonical paths.

This bit us on windows because of the \ vs / thing.
